File: c++_programming/all_programs/linkedlist_using_class.cpp
#include <iostream>
using namespace std;
class Node
{
	public:
	int data;
	Node *next;
	Node(int d)
	{
		this->data=d;
		this->next=NULL;
	}
};

class Linkedlist
{
	public:
	Node *head;
	Node *tail;
	Linkedlist()
	{
		this->head=NULL;
		this->tail=NULL;
	}

	void insert(int data)
	{
		Node *node=new Node(data);
		if(!this->head)
		{
			this->head=node;
		}
		else
		{
			this->tail->next=node;
		}
		this->tail=node;
	}
	Node* merge_sorted_list(Node *l1,Node *l2)
	{
		Node *current=NULL;
		Node *dummy_head=current;
		while(l1!=NULL && l2!=NULL)
		{
			if(l1->data<l2->data)
			{
				current->next=l1;
				l1=l1->next;
			}
			else if(l1->data>l2->data)
			{
				current->next=l2;
				l2=l2->next;
			}

			current=current->next;
		}
		if(l1)
			current->next=l1;
		else
			current->next=l2;

		return dummy_head->next;
	}

	void display()
	{
		Node *current=head;
		while(current)
		{
			cout<<current->data<<"\n"<<endl;
			current=current->next;
		}
	}
};

int main()
{
	Linkedlist *mylist=new Linkedlist();
	Linkedlist *slist=new Linkedlist();
	Linkedlist *tlist=new Linkedlist();
	slist->insert(5);
	slist->insert(11);
	mylist->insert(10);
	mylist->insert(20);
	tlist->head=tlist->merge_sorted_list(mylist->head,slist->head);
	tlist->display();
	return 0;
}

File: python/linkedlist_using_class.py
class Node:
    def __init__(self,data,next=None):
        self.data=data
        self.next=next

class Linkedlist:
    def __init__(self):
        self.head=None

    def add_node(self,data):
        newnode=Node(data)
        if self.head is None:
            self.head=newnode 
        else:
            n=self.head
            while n.next is not None:
                n=n.next
            n.next=newnode

    def print_node(self):
        curr=self.head
        while(curr):
            print(curr.data)
            curr=curr.next

    def merge_sorted_list(self,l1,l2):
        tail=dummy_head=Linkedlist()
        while l1 and l2:
            if l1.data<l2.data:
                tail.next,l1=l1,l1.next
            else:
                tail.next,l2=l2,l2.next
            
            tail=tail.next

        tail.next=l1 or l2
        return dummy_head.next

mylist=Linkedlist()
mylist.add_node(5)
mylist.add_node(10)
slist=Linkedlist()
slist.add_node(4)
slist.add_node(8)
#slist.print_node()
tlist=Linkedlist()
tlist.head=tlist.merge_sorted_list(mylist.head,slist.head)
tlist.print_node()
